### PR TITLE
Stop people from patching babylon by building it

### PR DESCRIPTION
**Pull request by sebmck**
Stops people doing evil stuff that we don't want to support or endorse.

### CODE REVIEW

File: packages/babylon/scripts/prepublish.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+require("./_util").updateMain("index.js");
+require("child_process").execSync(__dirname + "/../../../node_modules/.bin/browserify -s babylon -e " + __dirname + "/../lib/index.js -o " + __dirname + "/../index.js", { encoding: "utf8" });
```
Comment (sebmck) [inline packages/babylon/scripts/prepublish.js]: I don't need no stinkin node API.

**Comment by amasad** [comment]
what is an example of it being patched?

**Comment by sebmck** [comment]
https://github.com/shuhei/babel-plugin-angular2-annotations/blob/master/src/patch.js
https://bitbucket.org/rainydio/node-babel-plugin-thin-arrows/src/a58f223fa3e4b87d9740bef4b645a90c665d5ece/src/index.js?fileviewer=file-view-default

**Comment by amasad** [comment]
heh, ok. Can you add a comment somewhere explaining what's happening?
Afterwards feel free to merge. This needs to be a minor version bump?

**Comment by sebmck** [comment]
Sure. I'll leave this up for a day or two as it's not super urgent if anyone else wants to leave feedback. This should also make **slightly** faster since we wont be doing as much disk IO. Depending on how good this is we could consider doing it to more packages.

**Comment by forivall** [comment]
Can't we just be consenting adults, allow this, but nudge those who do to provide sufficient warning that they are doing this, and that users should expect no support from babel from this use?

**Comment by sebmck** [comment]
What they're doing is irresponsible and the fact that it's possible does not reflect well on Babel.

**Comment by forivall** [comment]
I think it reflects badly on babel that this change, imo, breaks the open/closed principle, by completely closing down the parser to extension.

**Comment by sebmck** [comment]
This isn't something that is supported, this has been very explicit and clear. I'm sorry you feel that way.

File: packages/babylon/scripts/prepublish.js (hunk 1 above)
Comment (samccone) [inline packages/babylon/scripts/prepublish.js]: path.join here  ... because windows

File: packages/babylon/scripts/postpublish.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+require("./_util").updateMain("lib/index.js");
+require("fs").unlinkSync(__dirname + "/../index.js");
```
Comment (samccone) [inline packages/babylon/scripts/postpublish.js]: path.joinnnnnn

File: packages/babylon/scripts/prepublish.js (hunk 1 above)
Comment (sebmck) [inline packages/babylon/scripts/prepublish.js]: ok mum

File: packages/babylon/scripts/_util.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+var fs = require("fs");
+
+exports.updateMain = function (main) {
+  var pkg = require("../package.json");
+  pkg.main = main;
+  fs.writeFileSync(__dirname + "/../package.json", JSON.stringify(pkg, null, "  "));
```
Comment (samccone) [inline packages/babylon/scripts/_util.js]: j o i n

Comment (sebmck) [inline packages/babylon/scripts/_util.js]: nope, this will work

File: packages/babylon/scripts/prepublish.js (hunk 1 above)
Comment (samccone) [inline packages/babylon/scripts/prepublish.js]: :baby: :baby_bottle:

**Comment by forivall** [comment]
I know it's never been supported. However, I do hope that it is something that could potentially be supported in the future.

Moreover, I don't completely understand why you think that this reflects badly on babel. Are there any pending bug reports resulting from people using these plugins? etc?

File: packages/babylon/scripts/_util.js (hunk 3 above)
Comment (samccone) [inline packages/babylon/scripts/_util.js]: :|

good :pencil2: http://shapeshed.com/writing-cross-platform-node/

:goat: :tangerine: :airplane: :earth_americas:

Comment (sebmck) [inline packages/babylon/scripts/_util.js]: :facepunch:

Comment (samccone) [inline packages/babylon/scripts/_util.js]: :bow: :cocktail:

Comment (sebmck) [inline packages/babylon/scripts/_util.js]: :no_entry_sign: :underage:

Comment (samccone) [inline packages/babylon/scripts/_util.js]: :older_man: :full_moon_with_face:

**Comment by thejameskyle** [comment]
Supporting users extending the parser would be a huge burden on the project if done improperly. And *if* we do want to do it someday, we would want to do so with careful consideration.

We're well aware how popular the idea of extending the Babel parser is, which is exactly why we're shutting it down before anything popular gets built on it and we're forced to support it. Because the truth is people are not as responsible as you give them credit for, and at the end of the day they are only placing a burden on us.

**Comment by sebmck** [comment]
Locking this as this isn't an appropriate place for this discussion and this change wont be reverted. Feel free to jump on the Babel Slack if you'd like to voice your concerns.

**Pull request merged**